probe-rs/src/architecture/arm/ap_v2/root_memory_interface.rs

```diff
@@ -0,0 +1,128 @@
+//! Functions to access the root memory interface of the Debug Port (DP) in an ADIv6
+//! implementation.
+use crate::{
+    architecture::arm::{
+        communication_interface::Initialized,
+        dp::{DpAccess, DpAddress, BASEPTR0, BASEPTR1},
+        memory::ArmMemoryInterface,
+        ApV2Address, ArmCommunicationInterface, ArmError, DapAccess, FullyQualifiedApAddress,
+    },
+    MemoryInterface,
+};
+
+/// The Root Memory Interface accesses the Debug Port (DP) address space. This memory interface can
+/// only be used to interface into the ROM tables and CoreSight components of the debug
+/// infrastructure.
+pub struct RootMemoryInterface<'iface> {
+    iface: &'iface mut ArmCommunicationInterface<Initialized>,
+    dp: DpAddress,
+    base: u64,
+}
+impl<'iface> RootMemoryInterface<'iface> {
+    pub fn new(
+        iface: &'iface mut ArmCommunicationInterface<Initialized>,
+        dp: DpAddress,
+    ) -> Result<Self, ArmError> {
+        let base_ptr0: BASEPTR0 = iface.read_dp_register(dp)?;
+        let base_ptr1: BASEPTR1 = iface.read_dp_register(dp)?;
+        let base = base_ptr0
+            .valid()
+            .then(|| u64::from(base_ptr1.ptr()) | u64::from(base_ptr0.ptr() << 12))
+            .inspect(|base| tracing::info!("DPv3 BASE_PTR: 0x{base:x}"))
+            .ok_or_else(|| ArmError::Other("DP has no valid base address defined.".into()))?;
+
+        Ok(Self { iface, dp, base })
+    }
+}
+
+impl MemoryInterface<ArmError> for RootMemoryInterface<'_> {
+    fn supports_native_64bit_access(&mut self) -> bool {
+        false
+    }
+
+    fn read_64(&mut self, _address: u64, _data: &mut [u64]) -> Result<(), ArmError> {
+        unimplemented!("The DPv3 only supports 32bit accesses")
+    }
+
+    fn read_32(&mut self, address: u64, data: &mut [u32]) -> Result<(), ArmError> {
+        //tracing::debug!("Reading {} words at {:x} on Root Access Port", data.len(), address);
+        // read content
+        for (i, d) in data.iter_mut().enumerate() {
+            let addr = address + (i as u64) * 4;
+            let base = (self.base + addr) & 0xFFFF_FFFF_FFFF_FFF0;
+            let fqa = FullyQualifiedApAddress::v2_with_dp(self.dp, ApV2Address::new_with_tip(base));
```

Comment:
I don't understand how the AP address is created here. 

That this is an implementation of `MemoryInterface` indicates to me that we can read from an arbitrary address, e.g. from a register in an AP. So `addr` in this context is not necessarily the base address of an AP.

And then `self.base` is used, which seems to come from the `BASEPTR0` and `BASEPTR1` registers. These point to the first component in the system, but I don't think this address should influence other accesses to APs.